### PR TITLE
Added compiler AST caching

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Added caching of the AST to the compiler 

This is a POC and should be considered carefully:

- do we need to have a toggle to enable/disable?
- is this the right place to put the cache files?

# Answers to questions

- Do we support removing the .cf so that you have compiled only files?  => No / not yet
- Why `__pycache__` => https://www.python.org/dev/peps/pep-3147/

# TODO

- [x] I would add it but disabled by default so that we can first test it a bit more
- [x] Maybe we need to force the pickle version so we do not run into issues there: https://docs.python.org/3/library/pickle.html#data-stream-format
- [ ] What happens if you compile with python 3.8 and run with 3.6?
- [x] The only difference between parse and my_parse is the caching. Perhaps naming them parse_cached and parse respectively would be better? Or even just parse as we had previously with an optional use_cache: bool parameter.
- [x] Move to mechanism similar to `__pycache__`, supporting multiple compiler versions

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
(docs added to config, params automatically, added note in the changelog)

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Comment by bartv** [comment]
Open questions for me:
- Do we support removing the .cf so that you have compiled only files?
- We need a mechanism to discard files when the compiler version is bumped
- I would add it but disabled by default so that we can first test it a bit more

**Review by bartv — CHANGES_REQUESTED** [review]
See my comments

**Comment by wouterdb** [comment]
get_complete_ast from from 31% to 8% of total time

image
image

File: tests/test_parser.py [hunk 1]
```diff
@@ -49,14 +49,14 @@
...
-from inmanta.parser.plyInmantaParser import parse
+from inmanta.parser.plyInmantaParser import myparse
 
 
 def parse_code(model_code: str):
     root_ns = Namespace("__root__")
     main_ns = Namespace("__config__")
     main_ns.parent = root_ns
-    statements = parse(main_ns, "test", model_code)
+    statements = myparse(main_ns, "test", model_code)
```
Comment (sanderr) [inline tests/test_parser.py:59]: This is a bit weird in my opinion. As I understand it the only difference between `parse` and `my_parse` is the caching. Perhaps naming them `parse_cached` and `parse` respectively would be better? Or even just `parse` as we had previously with an optional `use_cache: bool` parameter.

**Review by sanderr — CHANGES_REQUESTED** [review]
This is pretty cool.

File: tests/test_parser.py (hunk 1 above)
Comment (wouterdb) [reply tests/test_parser.py:59]: That would be better. I'll update it.

**Comment by sanderr** [comment]
To answer your questions:
- No strong opinion on the enable/disable
- I think it's as good a place as any to put the files. I haven't got much experience with this kind of stuff, but I think we could do worse than following Python on this

File: src/inmanta/parser/plyInmantaParser.py [hunk 2]
```diff
@@ -1040,7 +1042,25 @@ def myparse(ns: Namespace, tfile: str, content: Optional[str]) -> List[Statement
         return parser.parse(data, lexer=lexer, debug=False)
 
 
+def un_cache(namespace: Namespace, filename: str):
+    cache_filename = filename + "c"
+    if not os.path.exists(cache_filename):
+        return None
+    if os.path.getmtime(filename) > os.path.getmtime(cache_filename):
```
Comment (arnaudsjs) [inline src/inmanta/parser/plyInmantaParser.py:1049]:
```suggestion
    if os.path.getmtime(filename) != os.path.getmtime(cache_filename):
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]
* I would add a toggle
* I am not entirely convinced it's the right location to put the cached files. I am thinking about the scenario where you would like to remove all the cached files. If all the cached files are located in a certain directory that's easy. In the current model that isn't the case. Nobody is going to access the cached files manually, so there is no need to collocate them with the source files.

**Comment by bartv** [comment]
> * I am not entirely convinced it's the right location to put the cached files. I am thinking about the scenario where you would like to remove all the cached files. If all the cached files are located in a certain directory that's easy. In the current model that isn't the case. Nobody is going to access the cached files manually, so there is no need to collocate them with the source files.

If you store them somewhere else then cache management becomes more complex because you need a method to track to which file it belongs. An alternative is to use the new python method which is creating a directory in each dir for the cache files. Like for example __cache__

Removing them is also not that comples: `find -name *.cfc | xargs rm -f`

**Comment by arnaudsjs** [comment]
True. And any decent IDE has support to hide files of a certain type. Otherwise the project view will become very messy. Maybe that's something we should include in the Virtual studio code extension.

**Comment by bartv** [comment]
We should figure out why python did the move from pyc to `__pycache__`. I had not considered IDEs, that is probably the reason that python did this.

Cpython still supports both so that pre compiled files only works.

File: src/inmanta/parser/plyInmantaParser.py (hunk 2 above)
Comment (wouterdb) [reply src/inmanta/parser/plyInmantaParser.py:1049]: They will never be exactly the same (the original and the cached version are not created/modified at the same point in time)

File: changelogs/unreleased/ast-cache.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+description: Added experimental caching support to the compiler
+change-type: minor
+destination-branches: [master, iso4]
```
Comment (wouterdb) [inline changelogs/unreleased/ast-cache.yml:3]: @bartv are these the right places?

File: tests/test_parser.py (hunk 1 above)
Comment (wouterdb) [reply tests/test_parser.py:59]: (I didn't rename parse, because it is used in many places, if I can keep the API stable, I prefer that)

Comment (sanderr) [reply tests/test_parser.py:59]: I agree

Comment (wouterdb) [reply tests/test_parser.py]: but there may be a beter name for the underlying parse functionality

File: src/inmanta/app.py [hunk 4]
```diff
@@ -207,6 +207,44 @@ async def safe_shutdown_wrapper(shutdown_function: Callable[[], Coroutine[Any, A
         IOLoop.current().stop()
 
 
+class FeatureFlags:
```
Comment (sanderr) [inline src/inmanta/app.py:210]:
```suggestion
class ExperimentalFeatureFlags:
```
It looks like this class is solely for experimental features (implementation of `add_arguments`. The naming could make this clearer.

File: src/inmanta/parser/chache.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+"""
```
Comment (sanderr) [inline src/inmanta/parser/chache.py:1]: I didn't notice before but this file is named `chache.py` (with an additional "h").

File: src/inmanta/parser/chache.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+LOGGER = logging.getLogger(__name__)
+
+
+class CacheManager:
+    def __init__(self) -> None:
+        self.hits = 0
+        self.misses = 0
+        self.failures = 0
+
+    def get_file_name(self, filename: str) -> Optional[str]:
```
Comment (sanderr) [inline src/inmanta/parser/chache.py:37]: I think a docstring would be helpful here. The method name does not state sufficiently clearly what it does to go without.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/app.py (hunk 4 above)
Comment (wouterdb) [reply src/inmanta/app.py:210]: done

File: src/inmanta/app.py [hunk 7]
```diff
@@ -207,6 +207,44 @@ async def safe_shutdown_wrapper(shutdown_function: Callable[[], Coroutine[Any, A
...
+            parser.add_argument(
+                f"--experimental-{option.name}",
+                dest=metavar,
+                help=option.documentation,
+                action="store_true",
+                default=False,
+            )
+
+    def read_options_to_config(self, options: argparse.Namespace):
+        """ Read all feature flags from th argument parser into the configs v"""
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/app.py:237]: made an attempt to improve.

File: src/inmanta/app.py [hunk 8]
```diff
@@ -207,6 +207,44 @@ async def safe_shutdown_wrapper(shutdown_function: Callable[[], Coroutine[Any, A
...
+        for metavar, option in self.metavar_to_option.items():
+            value = getattr(options, metavar, False)
+            if value:
+                option.set("true")
+
+
+compiler_features = FeatureFlags()
+compiler_features.add(compiler.config.feature_compiler_cache)
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/app.py:245]: I had considered it, but the options would change names, which would be a problem on its own. 
So I chose instead to just wait until this naturally phases out.

File: src/inmanta/parser/chache.py (hunk 5 above)
Comment (wouterdb) [reply src/inmanta/parser/chache.py:1]: dyslexia (fixed)

File: src/inmanta/parser/chache.py (hunk 6 above)
Comment (wouterdb) [reply src/inmanta/parser/chache.py:37]: thx, added some comments

**Review by sanderr — APPROVED** [review]

File: src/inmanta/parser/cache.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                self.misses += 1
+                return None
+            if not os.path.exists(cache_filename):
+                self.misses += 1
+                return None
+            if os.path.getmtime(filename) > os.path.getmtime(cache_filename):
+                self.misses += 1
+                return None
+            with open(cache_filename, "rb") as fh:
+                self.hits += 1
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:76]: if  `ASTUnpickler()` raises an exception, the `hits` and the `failures` counter will be increased.

File: src/inmanta/parser/cache.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    def log_stats(self) -> None:
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return
+        LOGGER.info(
+            "Compiler cache observed %d hits and %d misses (%.d%%)",
+            self.hits,
+            self.misses,
+            (100 * self.hits) / (self.hits + self.misses),
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:107]: Shouldn't we also take into account the `failures` counter here?

Comment (arnaudsjs) [reply src/inmanta/parser/cache.py:107]: Should we reset the counters when `log_stats` is called?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/app.py [hunk 11]
```diff
@@ -252,6 +296,7 @@ def compiler_config(parser: ArgumentParser) -> None:
         action="store_true",
         default=False,
     )
+    compiler_features.add_arguments(parser)
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/app.py:299]: oh yes, good catch, thx.

File: src/inmanta/parser/cache.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        Returns the name for the cached file, based on the name of the original source file
+
+        Also ensures the cache folder exists.
+
+        :param filename: the filename of the source file
+        :return: the filename of the cached file
+        """
+        base_folder = os.path.dirname(filename)
+        if not os.path.isdir(base_folder):
+            return None
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/parser/cache.py]: this is indeed no longer required, the try except at the next level is a safer check.

File: src/inmanta/parser/cache.py (hunk 9 above)
Comment (wouterdb) [reply src/inmanta/parser/cache.py:80]: dang, good catch

File: src/inmanta/parser/cache.py (hunk 10 above)
Comment (wouterdb) [reply src/inmanta/parser/cache.py:111]: About failures: good question, to which I don't have an answer.

About resetting the counter: I'm opposed to that. 
If we also report these stats elsewhere (e.g. via the influxdb reporter), if every read resets, they will both get wrong data. 

In the current compiler, it is naturally reset, as it restarts every time. 

For other applications (language server, future compiler server), other aggregations or reset might be required.

File: src/inmanta/parser/cache.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+LOGGER = logging.getLogger(__name__)
+
+
+class CacheManager:
+    def __init__(self) -> None:
+        self.hits = 0
+        self.misses = 0
+        self.failures = 0
+
+    def get_file_name(self, filename: str) -> Optional[str]:
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:37]:
```suggestion
    def get_file_name(self, filename: str) -> str:
```

File: src/inmanta/parser/cache.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    def cache(self, filename: str, statements: List[Statement]) -> None:
+
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return
+        try:
+            cache_filename = self.get_file_name(filename)
+            if not cache_filename:
+                # could not find file
+                return
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:97]:
```suggestion
            cache_filename = self.get_file_name(filename)
```

File: src/inmanta/parser/cache.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+    def un_cache(self, namespace: Namespace, filename: str) -> Optional[List[Statement]]:
+        if not feature_compiler_cache.get():
+            # cache not enabled
+            return None
+        try:
+            cache_filename = self.get_file_name(filename)
+            if not cache_filename:
+                self.misses += 1
+                return None
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:72]:
```suggestion
            cache_filename = self.get_file_name(filename)
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

File: changelogs/unreleased/ast-cache.yml (hunk 3 above)
Comment (bartv) [reply changelogs/unreleased/ast-cache.yml:3]: yes

**Review by bartv — APPROVED** [review]

**Comment by sanderr** [comment]
@wouterdb this has been succesfully merged into iso4, you can manually merge it into master. Running the merge tool again would cause issues.

**Pull request merged**